ADBDEV-8057: Fix pg_rewind leaving replica on old timeline

silent-observer opened the pull request:
Remove bogus assertion in pg_rewind

It's entirely possible that the calculated "point of divergence" is
before the end of WAL on the target, because when we calculate the
point of divergence, we assume that the target's WAL extends to
infinity on its current timeline. It might be more clear to calculate
the end of WAL first, and if the target is a direct ancestor of the
source, consider the end of target WAL as the point of
divergence. However, it doesn't change the result: the target is a
direct ancestor of the source, and doesn't need to be rewound.

Add tests to cover the "no rewind required" cases, both when the
target's end of WAL is exactly at the point of divergence, which is
what the removed assertion assumed (case B in the test), and the case
where the assertion failed (case A).

Changes from original patch: test was made more general to prepare for more
detailed tests in the next commits. In particular, we pg_rewind from a
running source node, not from a stopped one, and the source node is passed as
an argument. This doesn't affect the test validity for this patch.

Discussion: https://www.postgresql.org/message-id/18575-1b9b3b60e2a480de@postgresql.org

---

Alternatively, perhaps just copying the history file would be enough?

Changes from original patch:
- Move sourceHistfile initialization to before rewind_parseTimeLineHistory,
  since rewind_parseTimeLineHistory actually modifies it.
- Don't copy timeline history file when dry_run is specified. Dry run
  shouldn't make changes that can potentially break the target server when
  applied incorrectly.
- Fix 107_empty_conf.pl test to preserve primary_conninfo setting.

---

Fix pg_rewind edge case when checkpoint is at switchpoint

In one specific edge case, when the source server makes a checkpoint exactly
after switching to a new timeline, switchpoint LSN exactly matches latest
checkpoint LSN. This can happen, for example, if the source has just performed
archive recovery. Then setting minRecoveryPoint to the checkpoint LSN in
pg_rewind is illegal and causes an error at startup, since minRecoveryPoint
actually means the minimum allowed end of WAL, and we want that to be after
the switch occurs, not at the end of the previous timeline.

As a fix, add +1 to the checkpoint LSN when setting minRecoveryPoint, which
now correctly handles this case. Also add a test scenario for the source node
that has performed recovery.

---

Note: Do not squash to preserve authorship

hlinnaka committed:
Remove bogus assertion in pg_rewind (#1970)

It's entirely possible that the calculated "point of divergence" is
before the end of WAL on the target, because when we calculate the
point of divergence, we assume that the target's WAL extends to
infinity on its current timeline. It might be more clear to calculate
the end of WAL first, and if the target is a direct ancestor of the
source, consider the end of target WAL as the point of
divergence. However, it doesn't change the result: the target is a
direct ancestor of the source, and doesn't need to be rewound.

Add tests to cover the "no rewind required" cases, both when the
target's end of WAL is exactly at the point of divergence, which is
what the removed assertion assumed (case B in the test), and the case
where the assertion failed (case A).

Changes from original patch: test was made more general to prepare for more
detailed tests in the next commits. In particular, we pg_rewind from a
running source node, not from a stopped one, and the source node is passed as
an argument. This doesn't affect the test validity for this patch.

Discussion: https://www.postgresql.org/message-id/18575-1b9b3b60e2a480de@postgresql.org

Ticket: ADBDEV-8057

RekGRpth commented on src/bin/pg_rewind/pg_rewind.c:
```diff
@@ -372,12 +375,39 @@ main(int argc, char **argv)
+				 */
+				open_target_file(sourceHistfileName, true);
+				write_target_range(sourceHistfile, 0, strlen(sourceHistfile));
+				close_target_file();
```
The original patch does this unconditionally.

silent-observer replied:
I think we shouldn't make potentially breaking changes to the target when dry run is enabled. Added explicit note to the description

KnightMurloc commented on src/bin/pg_rewind/t/010_target_is_ancestor.pl:
```diff
@@ -55,25 +55,30 @@
 $node_2->promote;
-$node_2->safe_psql('postgres', "INSERT INTO test_tab VALUES ('in node_2')");
-$node_2->stop;
+$node_2->safe_psql('postgres', "INSERT INTO test_tab VALUES ('in node')");
```
Why aren't these changes in the first commit?

silent-observer replied:
Moved most test changes to the first commit.

KnightMurloc commented:
> Use checkpoint+1 to account for the edge case when the checkpoint exactly
matches switchpoint LSN.

Is this not the same case as described in the first commit?

Add the appropriate description in the second commit.

hlinnaka committed:
Update minRecoveryPoint when no rewind is required (#1970)

Alternatively, perhaps just copying the history file would be enough?

Changes from original patch:
- Refactor sourceHistfile to be an out parameter instead of global state.
- Move sourceHistfile initialization to before rewind_parseTimeLineHistory,
  since rewind_parseTimeLineHistory actually modifies it.
- Don't copy timeline history file when dry_run is specified. Dry run
  shouldn't make changes that can potentially break the target server when
  applied incorrectly.
- Fix 107_empty_conf.pl test to preserve primary_conninfo setting.

Ticket: ADBDEV-8057

silent-observer committed:
Fix pg_rewind edge case when checkpoint is at switchpoint (#1970)

In one specific edge case, when the source server makes a checkpoint exactly
after switching to a new timeline, switchpoint LSN exactly matches latest
checkpoint LSN. This can happen, for example, if the source has just performed
archive recovery. Then setting minRecoveryPoint to the checkpoint LSN in
pg_rewind is illegal and causes an error at startup, since minRecoveryPoint
actually means the minimum allowed end of WAL, and we want that to be after
the switch occurs, not at the end of the previous timeline.

As a fix, add +1 to the checkpoint LSN when setting minRecoveryPoint, which
now correctly handles this case. Also add a test scenario for the source node
that has performed recovery.

Ticket: ADBDEV-8057

KnightMurloc commented on src/bin/pg_rewind/pg_rewind.c:
```diff
@@ -372,18 +375,38 @@ main(int argc, char **argv)
+
+			memcpy(&ControlFile_new, &ControlFile_target, sizeof(ControlFileData));
+			ControlFile_new.minRecoveryPoint = 	Max(ControlFile_source.minRecoveryPoint,
+													ControlFile_source.checkPoint + 1);
```
As I understand it, LSN should point to the beginning of the XLog record. If we just do +1, then LSN will point inside some record. How normal is this?

silent-observer replied:
`minRecoveryPoint` actually specifies the least acceptable end of WAL. If it points inside some record, that simply guarantees that the record will be processed. This is definitely okay for GPDB since `minRecoveryPoint` isn't used in that many places, and should probably be okay for raw Postgres too, although that'd require more research and discussion with upstream.

KnightMurloc replied:
Can you add some comment about this? Also, why can't we put `InvalidXLogRecPtr` here?

silent-observer replied:
`InvalidXLogRecPtr` would disable `minRecoveryPointTLI` as well, so it won't be guaranteed anymore that the standby will rewind to the switchpoint. If we want to use `minRecoveryPoint` functionality at all, we have to specify both of them. I will add the comment

KnightMurloc commented on src/bin/pg_rewind/pg_rewind.c:
```diff
@@ -372,18 +375,38 @@ main(int argc, char **argv)
 			rewind_needed = false;
+
+			memcpy(&ControlFile_new, &ControlFile_target, sizeof(ControlFileData));
+			ControlFile_new.minRecoveryPoint = 	Max(ControlFile_source.minRecoveryPoint,
+													ControlFile_source.checkPoint + 1);
```
Do we really need to update minRecoveryPoint? Is it not enough to update minRecoveryPointTLI?

silent-observer replied:
At startup there are checks that `minRecoveryPoint` is on the correct TLI, so no, we have to update them both. It's technically possible to not update the control file at all, but it's less robust and might fail in unexpected ways

KnightMurloc commented on src/bin/pg_rewind/pg_rewind.c:
```diff
@@ -372,18 +375,38 @@ main(int argc, char **argv)
 		{
-			/* the last common checkpoint record must be part of target WAL */
-			Assert(target_wal_endrec == divergerec);
+			ControlFileData ControlFile_new;
```
You're shadowing a variable. line 133

silent-observer replied:
Fixed

KnightMurloc commented on src/bin/pg_rewind/pg_rewind.c:
```diff
@@ -660,7 +683,11 @@ getTimelineHistory(ControlFileData *controlFile, int *nentries)
 		if (controlFile == &ControlFile_source)
+		{
 			histfile = fetchFile(path, NULL);
+			sourceHistfileName = pstrdup(path);
+			sourceHistfile = pstrdup(histfile);
```
I don't think it's a good idea to modify the global state from this function. Is there any way to put this logic into a separate function to get the contents of the history file?

silent-observer replied:
Unfortunately putting it into a separate function complicates the code a lot since the history file is also immediately processed in this function. I think it's easier to just pass them through out parameters instead.

KnightMurloc commented on src/bin/pg_rewind/t/107_empty_conf.pl:
```diff
@@ -31,6 +31,10 @@ sub run_test
 
 	RewindTest::run_pg_rewind($test_mode, do_not_start_primary => 1);
 
+	# Append new file to the old one to preserve primary_conninfo
+	my $new_file_data = slurp_file("$master_pgdata/postgresql.conf");
+	append_to_file("$tmp_folder/master-postgresql-full.conf.tmp", $new_file_data);
```
Why was this fix required?

silent-observer replied:
pg_rewind adds connection info to the config. Previously because minRecoveryPoint wasn't set and timeline files weren't copied, standby could finish recovery just from the already present WALs. This way it worked even with the original config, not modified by pg_rewind. However, after this patch the standby is now aware about the new timeline on primary and tried to retrieve it during recovery (which failed, because there is no primary_conninfo in the original config). So after this patch, we have to preserve changes made by pg_rewind, and not simply replace the config before starting the standby.

RekGRpth commented:
> Refactor sourceHistfile to be an out parameter instead of global state.

pgindent these your changes

silent-observer commented:
> Is it expected that changes in the src/bin/pg_rewind/t/107_empty_conf.pl test work without a patch?

Yes, changes to 107_empty_conf.pl only fix config line being accidentally deleted, they don't change the nature of the test. 010_target_is_ancestor.pl is the main test for this patch.
